Javascript filters for SAM and VCF

lindenb opened the pull request:
I wrote two related tools filtering SAM/VCF using ù*javascript** expressions and the JDK embedded javascript engine:

* src/java/picard/sam/SamFilterJavascript.java : is a picard version of my tool SAMJS https://github.com/lindenb/jvarkit/wiki/SamJS 

> Filters a SAM or BAM file with a javascript expression.  The script puts the following variables in the script context:  'record' a SamRecord ( https://samtools.github.io/htsjdk/javadoc/htsjdk/htsjdk/samtools/SAMRecord.html ) and  'header' a SAMFileHeader ( https://samtools.github.io/htsjdk/javadoc/htsjdk/htsjdk/samtools/SAMFileHeader.html ). Last value of the script should be a boolean to tell wether we should accept or reject the record.

```
INPUT=String
I=String                      The SAM or BAM file  Required. 

OUTPUT=File
O=File                        Output BAM file.  Required. 

JAVASCRIPT_FILE=File
F=File                        javascript file  Default value: null. 

JAVASCRIPT_EXPRESSION=String
E=String                      javascript expression  Default value: null. 
```

Example:

```
java -jar dist/picard.jar SamFilterJavascript I=testdata/picard/sam/aligned.sam O=jeter.sam E='record.getStart()%3==0'

```

src/java/picard/vcf/VcfFilterJavascript.java : is a picard version of my tool VCFFILTERJS https://github.com/lindenb/jvarkit/wiki/VCFFilterJS


> Filters a VCF file with a javascript expression.  The script puts the following variables in the script context:  'variant' a VariantContext ( https://samtools.github.io/htsjdk/javadoc/htsjdk/htsjdk/variant/variantcontext/VariantContext.html ) and  'header' a VCFHeader ( https://samtools.github.io/htsjdk/javadoc/htsjdk/htsjdk/variant/vcf/VCFHeader.html ). Last value of the script should be a boolean to tell wether we should accept or reject the record.


```
(...)INPUT=File
I=File                        Input file for analysis  Required. 

OUTPUT=File
O=File                        Output file.  Required. 

JAVASCRIPT_FILE=File
F=File                        javascript file  Default value: null. 

JAVASCRIPT_EXPRESSION=String
E=String                      javascript expression  Default value: null. 
```

example:

```
$ java -jar dist/picard.jar VcfFilterJavascript  I=testdata/picard/vcf/CEUTrio-snps.vcf O=jeter.vcf E='variant.isSNP() && variant.getStart()%2==0'
[Mon Nov 09 21:35:59 CET 2015] picard.vcf.VcfFilterJavascript INPUT=testdata/picard/vcf/CEUTrio-snps.vcf OUTPUT=jeter.vcf JAVASCRIPT_EXPRESSION=variant.isSNP() && variant.getStart()%2==0    VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false GA4GH_CLIENT_SECRETS=client_secrets.json
[Mon Nov 09 21:35:59 CET 2015] Executing as lindenb@bird on Linux 3.16.0-52-generic i386; Java HotSpot(TM) Server VM 1.8.0_51-b16; Picard version: 1.140(6a4a5d0088bfa4d84985465b966181596cf621b2_1447100655) JdkDeflater
INFO	2015-11-09 21:36:00	AbstractJavascriptFilterProgram	Compiling variant.isSNP() && variant.getStart()%2==0
[Mon Nov 09 21:36:00 CET 2015] picard.vcf.VcfFilterJavascript done. Elapsed time: 0.01 minutes.
Runtime.totalMemory()=64487424
```

I wrote some tests for both programs.

yfarjoun commented:
@lindenb I like the idea of having these filters, but I would like to run a different implementation architecture by you to see what you think. 

We already have some infrastructure for filtering both Sam records and VariantContexts, and we have an infrasturture for processing all the records in a sam-file (SinglePassSam). If you implement the abstract filter as a class that isn't a program, and extend it to both a SamRecordFilter and a VariantContextFilter, they could be used in other programs as filters (for example in a filtering iterator). Then you should be able to extend SinglePassSam and filter the reads you want using the filter. The VCF would be slightly more more, but you've already written most of what's needed. 

The advantage of this architecture would be 

1. less code duplication  (all your code for opening traversing a sam file) 
2. more useful filters (can be used in other programs)

What do you think of this alternative way?

lindenb commented:
> If you implement the abstract filter as a class that isn't a program, and extend it to both a SamRecordFilter and a VariantContextFilter, they could be used in other programs as filters (for example in a filtering iterator). 

ok, I'll be back later with this implementation

yfarjoun commented:
Thanks!

On Tue, Nov 10, 2015 at 2:34 AM, Pierre Lindenbaum <notifications@github.com
> wrote:

> If you implement the abstract filter as a class that isn't a program, and
> extend it to both a SamRecordFilter and a VariantContextFilter, they could
> be used in other programs as filters (for example in a filtering iterator).
>
> ok, I'll be back later with this implementation
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/broadinstitute/picard/pull/335#issuecomment-155345426>
> .
>

lbergelson commented:
Just have to say, this is pretty cool.

lindenb commented:
okay, 
* I removed my previous classes
* I created a new abstract file src/java/picard/util/JavascriptFilter.java . The classes contains what is needed for filtering with javascript: engine,compiled script and bindings. It contains two public static functions both returning concrete filters: VcfFilter and SamRecordFilter.
* I modified FilterVcf and FilterSamReads so they can use an instance of the previous filters
* I've added a test in src/tests/java/picard/vcf/TestFilterVcf.java (which was strangely assigned to the package picard.vcf.filter ? )

Sample ouputs:

SAM: print read unmapped or having start%2==0
```
~/src/picard$ java -jar dist/picard.jar FilterSamReads FILTER=includeJavascript INPUT=testdata/picard/sam/aligned.sam O=jeter.sam JS=<( echo "record.getReadUnmappedFlag() || record.getStart()%2==0;")

[Tue Nov 10 22:37:01 CET 2015] picard.sam.FilterSamReads INPUT=testdata/picard/sam/aligned.sam FILTER=includeJavascript [OUTPUT bam will contain reads that hava been accepted by the JAVASCRIPT_FILE script.] OUTPUT=jeter.sam JAVASCRIPT_FILE=/dev/fd/63    WRITE_READS_FILES=true VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false GA4GH_CLIENT_SECRETS=client_secrets.json
[Tue Nov 10 22:37:01 CET 2015] Executing as lindenb@bird on Linux 3.16.0-53-generic i386; Java HotSpot(TM) Server VM 1.8.0_51-b16; Picard version: 1.140(d55842e32a78d0305d7e9a91ddc19802f412ec7a_1447101057) JdkDeflater
INFO	2015-11-10 22:37:01	FilterSamReads	Filtering [presorted=true] aligned.sam -> OUTPUT=jeter.sam [sortorder=coordinate]
INFO	2015-11-10 22:37:01	FilterSamReads	4 SAMRecords written to jeter.sam
[Tue Nov 10 22:37:01 CET 2015] picard.sam.FilterSamReads done. Elapsed time: 0.01 minutes.
Runtime.totalMemory()=64487424

$ more jeter.sam 
@HD	VN:1.5	SO:coordinate
(...)
both_reads_align_clip_adapter	147	chr7	16	255	101M	=	21	-96	CAACAGAAGCNGGNATCTGTGTTTGTGTTTCGGA
TTTCCTGCTGAANNGNTTNTCGNNTCNNNNNNNNATCCCGATTTCNTTCCGCAGCTNACCTCCCAAN	)'.*.+2,))&&'&*/)-&*-)&.-)&)&),/-&&..)./.,.).*&&,&.&&-)&&&
0*&&&&&&&&/32/,01460&&/6/*0*/2/283//36868/&	RG:Z:0
both_reads_align_clip_marked	163	chr7	302	255	101M	=	1	-201	NCGCGGCATCNCGATTTCTTTCCGCAGCTAACCT
CCCGACAGATCGGCAGCGCGTCGTGTAGGTTATTATGGTACATCTTGTCGTGCGGCNAGAGCATACA	&/15445666651/566666553+2/14/&/555512+3/)-'/-&-'*+))*''13+
3)'//++''/'))/3+&*5++)&'2+&+/*&-&&*)&-./1'1	RG:Z:0
read_2_too_many_gaps	163	chr7	302	255	10M1D10M5I76M	=	1	-201	NCGCGGCATCNCGATTTCTTTCCGCAGCTAACCT
CCCGACAGATCGGCAGCGCGTCGTGTAGGTTATTATGGTACATCTTGTCGTGCGGCNAGAGCATACA	&/15445666651/566666553+2/14/&/555512+3/)-'/-&-'*+))*''13+
3)'//++''/'))/3+&*5++)&'2+&+/*&-&&*)&-./1'1	RG:Z:0
both_reads_present_only_first_aligns	165	*	0	0	*	chr7	1	0	NCGCGGCATCNCGATTTCTTTCCGCA
GCTAACCTCCCGACAGATCGGCAGCGCGTCGTGTAGGTTATTATGGTACATCTTGTCGTGCGGCNAGAGCATACA	&/15445666651/566666553+2/14/&/555512+3/)-'/-&-'*+
))*''13+3)'//++''/'))/3+&*5++)&'2+&+/*&-&&*)&-./1'1	RG:Z:0
```

VCF: flag variant where the REF allele is not 'A'
```
echo 'variant.getReference().getDisplayString().equals("A");' > refIsNotA.js
java -jar dist/picard.jar FilterVcf INPUT=testdata/picard/vcf/CEUTrio-snps.vcf O=jeter.vcf JS=refIsNotA.js

$ grep -v "#" jeter.vcf | cut -f 4,7 | sort | uniq
A	PASS
C	refIsNotA
G	refIsNotA
T	refIsNotA
```

yfarjoun commented on src/java/picard/sam/FilterSamReads.java:
```diff
@@ -110,7 +114,17 @@ public String toString() {
+	        + " The script puts the following variables in the script context: "
+	        + " 'record' a SamRecord ( https://samtools.github.io/htsjdk/javadoc/htsjdk/htsjdk/samtools/SAMRecord.html ) and "
+	        + " 'header' a SAMFileHeader ( https://samtools.github.io/htsjdk/javadoc/htsjdk/htsjdk/samtools/SAMFileHeader.html )."
+	        + " Last value of the script should be a boolean to tell wether we should accept or reject the record.",
```
...accept (true) or reject (false)...

yfarjoun commented on src/java/picard/sam/FilterSamReads.java:
```diff
@@ -168,24 +182,31 @@ protected int doWork() {
                             new ReadNameFilter(READ_LIST_FILE, false)));
                     break;
+                case includeJavascript:
+                	filterReads(new FilteringIterator(samReader.iterator(),
```
assert file is readable.

yfarjoun commented on src/java/picard/util/JavascriptFilter.java:
```diff
@@ -0,0 +1,215 @@
+/**
+ * Javascript filter with HEADER type containing TYPE records. contains two
+ * static method to get a SAM Read filter or a VariantFilter
+ *
```
@author?

yfarjoun commented on src/java/picard/util/JavascriptFilter.java:
```diff
@@ -0,0 +1,215 @@
+ *
+ */
+public abstract class JavascriptFilter<HEADER, TYPE> {
+	public static final String HEADER_KEY = "header";
```
rename DEFAULT_HEADER_KEY

yfarjoun commented:
@lindenb This is great!! 

a few overall comments.

1. I think you should use VariantContextFilter not VariantFilter, sorry about the confusing names.
2. Could you move the filters into htsjdk? That seems to be a better place for them. I'm happy to finish the review here so that it will be a fast-track in htsjdk. The CLPs should remain in Picard.

yfarjoun commented on src/java/picard/util/JavascriptFilter.java:
```diff
@@ -0,0 +1,215 @@
+	protected Bindings bindings;
+
+	/** put header in the bindings */
+	protected JavascriptFilter(final HEADER h) {
```
h->header

yfarjoun commented on src/java/picard/util/JavascriptFilter.java:
```diff
@@ -0,0 +1,215 @@
+				return false;
+			;
+			if (result instanceof Boolean) {
+				if (Boolean.FALSE.equals(result))
```
can you replace with return Boolean.TRUE.equals(result), with no additional if? seems clearer

yfarjoun commented on src/java/picard/util/JavascriptFilter.java:
```diff
@@ -0,0 +1,215 @@
+				if (Boolean.FALSE.equals(result))
+					return false;
+			} else if (result instanceof Number) {
+				if (((Number) result).intValue() != 1)
```
again, can you replace with return ((Number) result)==1, with no if statement?

yfarjoun commented on src/java/picard/util/JavascriptFilter.java:
```diff
@@ -0,0 +1,215 @@
+			} else {
+				return false;
+			}
+			return true;
```
if you did both things above, this line should be unneeded.

yfarjoun commented on src/java/picard/util/JavascriptFilter.java:
```diff
@@ -0,0 +1,215 @@
+		final ScriptEngineManager manager = new ScriptEngineManager();
+		this.engine = manager.getEngineByName("js");
+		if (this.engine == null) {
+			throw new RuntimeException("The embedded 'javascript' engine is not available in java. "
```
I'm uneasy about this, @lbergelson can you comment?

lbergelson replied:
I don't know very much about java javascript engines.  From what I can tell most jvms should include some js engine, either rhino or nashorn.  Things like android might be an exception, but I think anyone running picard on android can be expected to put a bit of work in for compatibility.    I think we should make this exception more specific so someone could attempt to catch it should they want to have different error reporting.  Could you add a new subclass of `PicardException` called `JavaScriptException` and throw that instead.

yfarjoun commented on src/java/picard/util/JavascriptFilter.java:
```diff
@@ -0,0 +1,215 @@
+		} finally {
+			CloserUtil.close(scriptReader);
+		}
+
```
spurious empty line.

yfarjoun commented on src/java/picard/util/JavascriptFilter.java:
```diff
@@ -0,0 +1,215 @@
+
+	/**
+	 * creates a SamRecordFilter from the given SAMFileHeader and the user
+	 * javascript file
```
expand javadoc to include @param arguments and @return value

yfarjoun commented on src/java/picard/util/JavascriptFilter.java:
```diff
@@ -0,0 +1,215 @@
+	 * creates a VariantFilter from the given VCFHeader and the user javascript
+	 * file
+	 */
+	static public VariantFilter makeVariantContextFilter(final VCFHeader header, final File scriptFile) {
```
VariantContextFilter

yfarjoun replied:
also expand javadoc

yfarjoun commented on src/java/picard/util/JavascriptFilter.java:
```diff
@@ -0,0 +1,215 @@
+
+		@Override
+		public boolean filterOut(final SAMRecord first, final SAMRecord second) {
+			return filterOut(first) || filterOut(second);
```
due to the reverse logic, this will exclude read-pairs if either one of them gets false from the javascript. This should either be an option, or not allowed, or clearly stated, since it isn't obvious from reading.

yfarjoun commented on src/java/picard/util/JavascriptFilter.java:
```diff
@@ -0,0 +1,215 @@
+			return !accept(record);
+		}
+
+		@Override
```
no need for this, please remove

yfarjoun replied:
oops. sorry, this is needed.

yfarjoun commented on src/java/picard/util/JavascriptFilter.java:
```diff
@@ -0,0 +1,215 @@
+		}
+
+		/** get filter name , use the scriptfile filename, remove the suffix */
+		public String getFilterName() {
```
not needed (if moving to VariantContextFilter)

yfarjoun commented on src/java/picard/util/JavascriptFilter.java:
```diff
@@ -0,0 +1,215 @@
+		}
+
+		@Override
+		public List<VCFFilterHeaderLine> headerLines() {
```
not needed.

yfarjoun commented on src/java/picard/vcf/filter/FilterVcf.java:
```diff
@@ -20,6 +20,7 @@
+ * 
  */
 package picard.vcf.filter;
 
```
My sense is that this is the wrong file for this. This file modifies VCFs FILTER field, not what you want, right? I'm assuming you want to actually remove records from the VCF. Please correct me if I'm wrong about this.

lindenb replied:
(Thank you for the whole review )
That's why I've implemented the javascript filter as a VariantFilter (and not VariantContextFilter ) to be used with **FilterVcf** .
For VCF, as far as I can see, there is no existing instance of CommandLineProgram that could be used. Are you OK with creating a new subclass of CommandLineProgram , say *VcfFilterVariants* ?

yfarjoun replied:
Yes, know there's not such program, I would be very happy if you wrote one.

yfarjoun commented on src/tests/java/picard/vcf/TestFilterVcf.java:
```diff
@@ -41,7 +43,33 @@
+			throw new PicardException("Return value non-zero: " + retval);
+		}
+	}
+    
```
please test the samfilter as well. also, it would be nice to test that it actually worked...as in, did the thing you wanted, not just completed successfully.

lindenb replied:
* the enum Filter of FilterSamReads is private
* doWork is protected

are you ok with moving those to 'public' in order to create a new test (There is no existing test for FilterSamReads ) ?

yfarjoun replied:
feel free to move the Enum to protected so that the test can use it.
if the test is in the same package, it should be able to use it.

yfarjoun commented:
back you you @lindenb

lbergelson commented on src/java/picard/util/JavascriptFilter.java:
```diff
@@ -0,0 +1,215 @@
+			        + "Do you use the SUN/Oracle Java Runtime ?");
+		}
+		if (scriptFile == null)
+			throw new NullPointerException("missing javascript file.");
```
Please throw `IllegalArgumentException` here instead of null pointer.  Also, always use `{}` in your if statements even for one liners.

lbergelson commented on src/java/picard/util/JavascriptFilter.java:
```diff
@@ -0,0 +1,215 @@
+			scriptReader = new FileReader(scriptFile);
+			this.script = compilingEngine.compile(scriptReader);
+		} catch (IOException err) {
+			throw new RuntimeException("I/O exception in " + scriptFile, err);
```
Use `RuntimeIOException` instead.

lbergelson commented on src/java/picard/util/JavascriptFilter.java:
```diff
@@ -0,0 +1,215 @@
+		} catch (IOException err) {
+			throw new RuntimeException("I/O exception in " + scriptFile, err);
+		} catch (ScriptException err) {
+			throw new RuntimeException("Script error in " + scriptFile, err);
```
Throw the new `JavaScriptException` instead of runtime.

lbergelson commented on src/java/picard/util/JavascriptFilter.java:
```diff
@@ -0,0 +1,215 @@
+			}
+			return true;
+		} catch (ScriptException err) {
+			throw new RuntimeException(err);
```
Use `JavaScriptException` as mentioned below

lbergelson commented on src/java/picard/util/JavascriptFilter.java:
```diff
@@ -0,0 +1,215 @@
+	private CompiledScript script = null;
+
+	/** javascript engine */
+	private ScriptEngine engine = null;
```
this shouldn't be a field since it's only used in initialize, move it to there

lbergelson commented on src/java/picard/util/JavascriptFilter.java:
```diff
@@ -0,0 +1,215 @@
+			throw new NullPointerException("missing javascript file.");
+		IOUtil.assertFileIsReadable(scriptFile);
+		try {
+			final Compilable compilingEngine = (Compilable) this.engine;
```
This cast makes me worried.  Is it the case that all javascript engines are `Compilable`? It looks like Rhino and Nashorn both are.  Are there any others?

Could you also extract this and the above engine getting code into it's own static function `getJavaScriptEngine` that returns a Compileable.

lbergelson commented on src/java/picard/util/JavascriptFilter.java:
```diff
@@ -0,0 +1,215 @@
+		/** get filter name , use the scriptfile filename, remove the suffix */
+		public String getFilterName() {
+			String fname = this.scriptFile.getName();
+			if (fname.toLowerCase().endsWith(".js")) {
```
use `IOUtil.baseName` instead of this if-else

lbergelson commented on src/java/picard/util/JavascriptFilter.java:
```diff
@@ -0,0 +1,215 @@
+public abstract class JavascriptFilter<HEADER, TYPE> {
+	public static final String HEADER_KEY = "header";
+	/** compiled user script */
+	private CompiledScript script = null;
```
It's my understanding that you won't ever reinitialize one of these.  Could you make script be private and final, and replace initialize with a static method called "compileScript" which is called in constructor and returns a compiledScript.

lbergelson commented on src/tests/java/picard/vcf/TestFilterVcf.java:
```diff
@@ -41,7 +43,33 @@
+
+		final int retval = filterer.doWork();
+		if (retval != 0) {
+			throw new PicardException("Return value non-zero: " + retval);
```
Instead of 
```
if (retval != 0) {
			throw new PicardException("Return value non-zero: " + retval);
}
```
it's more idiomatic to use
`Assert.assertEquals(retval, 0)`

lbergelson commented:
@lindenb As I said before, this is very cool.  Thank you for contributing this code.  I had a few refactoring/ stylistic suggestions.

lindenb commented:
@lbergelson thanks for the comment and the review, I'll start refactoring my code when I'll get some answers from @yfarjoun about the tests and what's would be the best implementation of a js-filter for vcf (there is no hurry :-) )

lindenb commented:
@yfarjoun , for a start, as you requested, I've moved the filters to htsjdk: https://github.com/samtools/htsjdk/pull/385

lindenb commented:
@yfarjoun FYI: the pull request for htsjdk has been accepted https://github.com/samtools/htsjdk/pull/385 , I'm going to wait for the next release of htsjdk before going bask to this PR.

yfarjoun commented:
great! thanks for your patience.

On Thu, Dec 17, 2015 at 3:20 PM, Pierre Lindenbaum <notifications@github.com
> wrote:

> @yfarjoun <https://github.com/yfarjoun> FYI: the pull request for htsjdk
> has been accepted samtools/htsjdk#385
> <https://github.com/samtools/htsjdk/pull/385> , I'm going to wait for the
> next release of htsjdk before going bask to this PR.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/broadinstitute/picard/pull/335#issuecomment-165571794>
> .
>

yfarjoun commented:
@lindenb just checking that the ball is in your court here...correct?

lindenb commented:
@yfarjoun I'll update this PR for the next version of htsjdk which should contain those JS filters. I can close this PR and submit later if you want.

yfarjoun commented:
doesn't bother me, just making sure you're not waiting on a review.

On Sun, Jan 10, 2016 at 5:07 PM, Pierre Lindenbaum <notifications@github.com
> wrote:

> @yfarjoun <https://github.com/yfarjoun> I'll update this PR for the next
> version of htsjdk which should contain those JS filters. I can close this
> PR and submit later if you want.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/broadinstitute/picard/pull/335#issuecomment-170398460>
> .
>

lindenb commented:
@yfarjoun I've got time to work on this: Is it possible to work on picard/master using the current htsjdk/master (instead of the latest htsjdk release ?) without breaking the tests ? . How should I do this ?

yfarjoun commented:
Not sure I understand your question. The automatic tests here (Travis) will always pull the latest master branch of htsjdk (not latest release), and you can put whatever you like in your own working directory...in particular you can put the latest master.

lindenb commented:
> The automatic tests here (Travis) will always pull the latest master branch of htsjdk (not latest release)

that's what I wanted to know
thanks.

lindenb commented:
@yfarjoun @lbergelson I'm back to this PR now that some javascript filters have been included in htsjdk.

- I use those filters in picard
- I had to create a extends the variantContextFilter so it implements picard/VariantFilter
- I created some tests for sam and vcf. I move the enum of FilterSam to public.
- strangely I had to change the package of https://github.com/lindenb/picard/blob/5e8bcf483da749fa227b6d2515d3261d05156bd9/src/tests/java/picard/vcf/TestFilterVcf.java (wrong package ?)

lindenb commented:
anyone ?

lbergelson commented:
Hey @lindenb, I think Yossi and I both figured the other was taking a look at this.

lbergelson commented on src/java/picard/sam/FilterSamReads.java:
```diff
@@ -20,6 +20,9 @@
  * THE SOFTWARE.
+ * 
+ * History:
+ *  2015-11: Pierre Lindenbaum added the javascript filter
```
In general we typically just let git blame keep track of the history.  No need to delete it, but since it's not usually updated it's not usually a reliable source of information.

lbergelson commented on src/java/picard/sam/FilterSamReads.java:
```diff
@@ -79,12 +83,12 @@
     private static final Log log = Log.getInstance(FilterSamReads.class);
 
-    private static enum Filter {
+    public static enum Filter {
```
Why did this change to be public?

lbergelson replied:
Please revert it to private if you can (looks like you can).  Also, could you remove the static keyword while you're at it?  It's redundant.

lbergelson commented on src/java/picard/sam/FilterSamReads.java:
```diff
@@ -181,24 +195,33 @@ protected int doWork() {
             switch (FILTER) {
                 case includeAligned:
-                    filterReads(new FilteringIterator(SamReaderFactory.makeDefault().referenceSequence(REFERENCE_SEQUENCE).open(INPUT).iterator(),
+                    filterReads(new FilteringIterator(samReader.iterator(),
```
+1 to extracting that long repeated line into a variable, but why not go all the way and move the iterator creation to the same place?

lbergelson commented on src/java/picard/vcf/filter/AlleleBalanceFilter.java:
```diff
@@ -64,7 +64,7 @@ public String filter(final VariantContext ctx) {
 
         for (final Genotype gt : ctx.getGenotypesOrderedByName()) {
-            if (gt.isNoCall() || !gt.isHet()) continue;
+            if (gt.isNoCall() || !gt.isHet() || !gt.hasAD()) continue;
```
good catch

lbergelson commented on src/java/picard/vcf/filter/FilterVcf.java:
```diff
@@ -87,14 +103,26 @@ public static void main(final String[] args) {
+            try {
+                variantFilters.add(new VariantContextJavascriptFilter(JAVASCRIPT_FILE, in.getFileHeader()));
+            } catch(IOException error) {
+                CloserUtil.close(in);
```
I appreciate that you're closing the input stream when necessary.  Would it be easier to wrap the whole thing in a try block and instantiate the input reader in the try-initialization?

lbergelson commented on src/java/picard/vcf/filter/FilterVcf.java:
```diff
@@ -87,14 +103,26 @@ public static void main(final String[] args) {
+        if( JAVASCRIPT_FILE != null) {
+            try {
+                variantFilters.add(new VariantContextJavascriptFilter(JAVASCRIPT_FILE, in.getFileHeader()));
+            } catch(IOException error) {
```
error can be final.  (you'll notice you get a lot of these comments from reviewers.  It's because we all use intellij which autodetects what can be made final and pops up a little yellow bar next to it if it's not)

lbergelson commented on src/java/picard/vcf/filter/FilterVcf.java:
```diff
@@ -131,4 +160,42 @@ private boolean isVcfOrBcf(final File file) {
+        private final String filterName;
+        /** script file */
+        private final File scriptFile;
+        private VariantContextJavascriptFilter(final File scriptFile, final VCFHeader header) throws IOException {
```
probably want another newline between scriptFile and here

lbergelson commented on src/java/picard/vcf/filter/FilterVcf.java:
```diff
@@ -87,14 +103,26 @@ public static void main(final String[] args) {
 
     @Override
-    protected int doWork() {
+    public int doWork() {
```
Did this have to change to be public?  I think the test should be moved into the same package and this be left as protected.  It's strange for a subclass to increase the visibility of a function.

lbergelson commented on src/tests/java/picard/sam/FilterSamReadsTest.java:
```diff
@@ -0,0 +1,62 @@
+        // input as SAM file 
+        final File inputSam = new File("testdata/picard/sam/aligned.sam");
+        //loop over javascript filters
+        for(final File javascriptFile: new File("testdata/picard/sam/FilterSamReads").listFiles(new FileFilter() {
```
This auto discovery is cool, I think it would be better to just hardcode the names though.  This would be a good place to use a dataprovider.  It would be good to test if the expected number of reads were actually filtered out as well.

lbergelson commented on src/java/picard/vcf/filter/FilterVcf.java:
```diff
@@ -131,4 +160,42 @@ private boolean isVcfOrBcf(final File file) {
+           this.scriptFile = scriptFile;
+           /* create filter name using file basename */
+           String fname = scriptFile.getName();
+            if (fname.toLowerCase().endsWith(".js")) {
```
This might be written more cleanly using `IOUtil.basename` and `fileSuffix`

lbergelson commented on src/tests/java/picard/vcf/TestFilterVcf.java:
```diff
@@ -43,6 +45,31 @@
+		pw.println(content);
+		pw.flush();
+		pw.close();
+		return out;
```
I think you could save a few lines with

```
try(PrintWriter pw = new PrintWriter(out){
  pw.println(content)
}
```

the close and flush should be called implicitly that way.

lbergelson commented on src/tests/java/picard/vcf/TestFilterVcf.java:
```diff
@@ -43,6 +45,31 @@
+
+	@Test
+	public void testJavaScript() throws Exception {
+		final File out = File.createTempFile("filterVcfTestJS.", ".vcf");
```
It would be good if this checked that the correct variants were actually filtered.

lbergelson commented on testdata/picard/sam/FilterSamReads/filtersamreads01.js:
```diff
@@ -0,0 +1,2 @@
+/** reads mapped and having start%2 = 0 */
```
could you give these filters slightly more descriptive names?

maybe `filterOddStarts.js` and `filterReadsWithout5primeSoftClip`

lbergelson commented:
@lindenb I did another pass on this.  You're probably sick of me and the try-with-resources comments by now.  Ping me when you're ready and it will be good to merge.  

Thanks for this work.  It's very cool stuff. I'm sorry we're not always quick to respond to things.  Lots of stuff going on internally here and not enough reviewers.  Don't be afraid to ping us like you did if you feel like something is being ignored.

lindenb commented:
@lbergelson back to you @lbergelson 

note: a major change was to move : src/tests/java/picard/vcf/TestFilterVcf.java -> src/tests/java/picard/vcf/filter/TestFilterVcf.java so I can use the fields of FilterVcf

* https://github.com/broadinstitute/picard/pull/335#discussion_r53528183 remove 'history paragraph from the header of FilterSamReaders"
* https://github.com/broadinstitute/picard/pull/335#discussion_r53528805 the enum 'Filter' is moved to 'protected' so we can use it in FilteSamReadsTest
* https://github.com/broadinstitute/picard/pull/335#discussion_r53528903 removed static from enum
* https://github.com/broadinstitute/picard/pull/335#discussion_r53529112 moved Iterator out of the switch/case 
* https://github.com/broadinstitute/picard/pull/335#discussion_r53529682 wrapping the '*.close()' in a big try/catch. Moved 'in' and 'out' out of the try/catch
* https://github.com/broadinstitute/picard/pull/335#discussion_r53529880 error is final
* https://github.com/broadinstitute/picard/pull/335#discussion_r53530496 newline added
* https://github.com/broadinstitute/picard/pull/335#discussion_r53530857 problem with scope doWork because TestFilterVcf is in the wrong package. moving TestFilterVcf from picard.vcf to picard.vcf.filter
* https://github.com/broadinstitute/picard/pull/335#discussion_r53530905 added auto-discovery annotation
* https://github.com/broadinstitute/picard/pull/335#discussion_r53531744 using IOUtils.basename
* https://github.com/broadinstitute/picard/pull/335#discussion_r53532230 wrapping printwrite into 'try-with-rsrc' statement.
* https://github.com/broadinstitute/picard/pull/335#discussion_r53532262 checking the number of variants in the vcf test
* https://github.com/broadinstitute/picard/pull/335#discussion-diff-53532849 changed the jsfilter as you suggested

lbergelson commented:
@lindenb  :+1: 
Thanks for addressing comments.
